src/MongoDB/BulkWrite.c

```diff
@@ -73,6 +73,70 @@ static inline bool php_phongo_bulkwrite_update_has_operators(bson_t* bupdate) /*
 	return false;
 } /* }}} */
 
+/* Returns whether the BSON array's keys are a sequence of integer strings
+ * starting with "0". BSON_APPEND_ARRAY considers it the caller's responsibility
+ * to ensure that the array's keys are properly formatted. */
+static inline bool php_phongo_bulkwrite_bson_array_has_valid_keys(bson_t* array) /* {{{ */
+{
+	bson_iter_t iter;
+
+	if (bson_empty(array)) {
+		return true;
+	}
+
+	if (bson_iter_init(&iter, array)) {
+		char key[12];
+		int  count = 0;
+
+		while (bson_iter_next(&iter)) {
+			bson_snprintf(key, sizeof(key), "%d", count);
+
+			if (0 != strcmp(key, bson_iter_key(&iter))) {
+				return false;
+			}
+
+			count++;
+		}
+	}
+
+	return true;
+} /* }}} */
+
+/* Appends an array field for the given opts document and key. Returns true on
+ * success; otherwise, false is returned and an exception is thrown. */
+static bool php_phongo_bulkwrite_opts_append_array(bson_t* opts, const char* opts_key, zval* zarr, const char* zarr_key TSRMLS_DC) /* {{{ */
+{
+	zval*  value = php_array_fetch(zarr, zarr_key);
+	bson_t b     = BSON_INITIALIZER;
+
+	if (Z_TYPE_P(value) != IS_OBJECT && Z_TYPE_P(value) != IS_ARRAY) {
+		phongo_throw_exception(PHONGO_ERROR_INVALID_ARGUMENT TSRMLS_CC, "Expected \"%s\" option to be array or object, %s given", zarr_key, zend_get_type_by_const(Z_TYPE_P(value)));
+		return false;
+	}
+
+	php_phongo_zval_to_bson(value, PHONGO_BSON_NONE, &b, NULL TSRMLS_CC);
+
+	if (EG(exception)) {
+		bson_destroy(&b);
+		return false;
+	}
+
+	if (!php_phongo_bulkwrite_bson_array_has_valid_keys(&b)) {
+		phongo_throw_exception(PHONGO_ERROR_INVALID_ARGUMENT TSRMLS_CC, "\"%s\" option has invalid keys for a BSON array", zarr_key, zend_get_type_by_const(Z_TYPE_P(value)));
```

Comment:
This has an extra ``zend_get_type_by_const`` — Travis says so too: https://travis-ci.org/mongodb/mongo-php-driver/jobs/383741731#L1198